### PR TITLE
refactor: Refactor of attribute lookahead processing

### PR DESCRIPTION
**Pull request by asmelko**
Until the PR, attribute lookahead was triggered during evaluation of CA expressions. That caused a complexity on the parser as it needed to be prepared to start parsing (performing lookup) in the middle of any statement processing.

The PR introduces new method of performing attribute lookup. It separates lookup from statement processing by initiating the lookup before a statement is passed to a processor.

The refactor caused the simplification of `parser_impl` and `collector` classes. Also it moved the functionality of lookup from `processing_manager` to `statement_provider` while utilizing the `processing_manager`'s loop.  

The refactor may simplify the future work in parsing optimizations.

### CODE REVIEW

File: parser_library/src/parsing/grammar/operand_field_rules.g4 [hunk 1]
```diff
@@ -283,27 +291,17 @@ op_rem_body_deferred
 	: 
 	{
 		collector.set_operand_remark_field(provider.get_empty_range(_localctx->getStart()));
-						collector.add_operands_hl_symbols();
-						collector.add_remarks_hl_symbols();
-						process_statement();
 	} EOLLN EOF
-	| SPACE {enable_hidden();} deferred_op_rem 
+	| SPACE {enable_hidden();} deferred_op_rem {disable_hidden();}
```
[the remark this answers is not in the document]
Comment (asmelko) [reply parser_library/src/parsing/grammar/operand_field_rules.g4:295]: I think that ANTLR parser does not just tries rules if they match. It has pre-built decision map.

File: parser_library/src/processing/processing_manager.cpp [hunk 2]
```diff
@@ -25,15 +25,13 @@
...
 
-using namespace hlasm_plugin::parser_library;
-using namespace hlasm_plugin::parser_library::processing;
-using namespace hlasm_plugin::parser_library::workspaces;
+namespace hlasm_plugin::parser_library::processing {
 
 processing_manager::processing_manager(std::unique_ptr<opencode_provider> base_provider,
     context::hlasm_context& hlasm_ctx,
-    const library_data data,
+    const workspaces::library_data data,
```
[the remark this answers is not in the document]
Comment (asmelko) [reply parser_library/src/processing/processing_manager.cpp:32]: no, lines 54 and 59 use the other field of `data`

File: parser_library/src/processing/context_manager.cpp [hunk 3]
```diff
@@ -98,12 +98,15 @@ context::id_index context_manager::get_symbol_name(const std::string& symbol, ra
 
 context_manager::name_result context_manager::try_get_symbol_name(const std::string& symbol) const
 {
+    if (symbol.empty() || symbol.size() > 63)
+        return std::make_pair(false, context::id_storage::empty_id);
+
     bool is_symbol = true;
     for (size_t i = 0; i < symbol.size(); ++i)
         if (!lexing::lexer::ord_char(symbol[i]) || !(i != 0 || !isdigit(symbol[i])))
             is_symbol = false;
```
Comment (slavek-kucera) [inline parser_library/src/processing/context_manager.cpp]: the following seems more readable...
```suggestion
    for (size_t i = 0; i < symbol.size(); ++i)
        if (!lexing::lexer::ord_char(symbol[i]) || i == 0 && isdigit(symbol[i]))
            return std::make_pair(false, context::id_storage::empty_id);
```
or maybe even
```suggestion
    if (isdigit(symbol.front()))
        return std::make_pair(false, context::id_storage::empty_id);
    for (auto s : symbol)
        if (!lexing::lexer::ord_char(s))
            return std::make_pair(false, context::id_storage::empty_id);
```

File: parser_library/src/context/ordinary_assembly/ordinary_assembly_context.cpp [hunk 4]
```diff
@@ -55,6 +54,15 @@ bool ordinary_assembly_context::create_symbol(
     return ok;
 }
 
+void ordinary_assembly_context::add_symbol_reference(symbol sym) { symbol_refs_.try_emplace(sym.name, std::move(sym)); }
```
[the remark this answers is not in the document]
Comment (slavek-kucera) [reply parser_library/src/context/ordinary_assembly/ordinary_assembly_context.cpp:57]: This is actually a very good question. The move itself is not an issue - that's just a cast.
In the case of key already being in the map, nothing happens.
If a new node is to be added, I think that the new value pair will be constructed in the following order - copy the key into `.first` and then move the whole sym into `.second`.

File: parser_library/src/processing/processing_manager.cpp [hunk 5]
```diff
@@ -119,8 +117,22 @@ void processing_manager::start_processing(std::atomic<bool>* cancel)
...
+{
+    auto look_pr = dynamic_cast<lookahead_processor*>(&*procs_.back());
+    return look_pr && look_pr->action == lookahead_action::ORD;
+}
+
 statement_provider& processing_manager::find_provider()
 {
+    if (attr_lookahead_active())
+    {
+        auto& opencode_prov = **(provs_.end() - 1);
```
[the remark this answers is not in the document]
Comment (asmelko) [reply parser_library/src/processing/processing_manager.cpp:130]: yes, opencode and copy provs are always created in the constructor

File: parser_library/src/context/ordinary_assembly/ordinary_assembly_context.cpp [hunk 6]
```diff
@@ -54,7 +54,11 @@ bool ordinary_assembly_context::create_symbol(
     return ok;
 }
 
-void ordinary_assembly_context::add_symbol_reference(symbol sym) { symbol_refs_.try_emplace(sym.name, std::move(sym)); }
+void ordinary_assembly_context::add_symbol_reference(symbol sym)
+{
+    auto name = sym.name;
+    symbol_refs_.try_emplace(name, std::move(sym));
```
Comment (slavek-kucera) [inline parser_library/src/context/ordinary_assembly/ordinary_assembly_context.cpp:60]: I still think that the copy is not needed. But this code will make two copies...

**Pull request merged**